src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp

```diff
@@ -1718,7 +1718,11 @@ class ShenandoahGenerationalEvacuationTask : public WorkerTask {
           // doing this work during a safepoint.  We cannot put humongous regions into the collection set because that
           // triggers the load-reference barrier (LRB) to copy on reference fetch.
           r->promote_humongous();
-        } else if (r->is_regular() && (r->garbage_before_padded_for_promote() < old_garbage_threshold) && (r->get_top_before_promote() == tams)) {
+        } else if (r->is_regular() && (r->get_top_before_promote() != nullptr)) {
+          assert(r->garbage_before_padded_for_promote() < old_garbage_threshold,
+                 "Region " SIZE_FORMAT " has too much garbage for promotion", r->index());
+          assert(r->get_top_before_promote() == tams,
```

Comment:
Where was this second condition established when identifying a region as eligible for promotion-in-place? As far as I can see the second condition is tested _after_ the top has been saved.

Reply:
That's a good catch. We do check this condition, but we do it after we set the sentinel that indicates this region should be promoted. I'll fix this.